metals/src/main/scala/scala/meta/internal/metals/MetalsEnrichments.scala

```diff
@@ -1077,4 +1077,17 @@ object MetalsEnrichments
     def toLsp = new l.Position(breakpoint.getLine() - 1, breakpoint.getColumn())
   }
 
+  implicit class XtensionWorkspaceEdits(edits: Seq[l.WorkspaceEdit]) {
+    def mergeChanges: l.WorkspaceEdit = {
+      val changes = edits.view
+        .flatMap(e => Option(e.getChanges()).map(_.asScala))
+        .flatten
+        .groupMapReduce(_._1)(_._2.asScala) { _ ++ _ }
+        .mapValues(_.toSet.toSeq.asJava) // deduplicate same changes
+        .toMap
+        .asJava
+      new l.WorkspaceEdit(changes)
```

Comment:
A bit more verbose, but also slightly easier to digest (I hope)
```suggestion
      val changesPerDocument = edits.view
        .flatMap(edit => Option(edit.getChanges).map(_.asScala))
        .flatMap(_.toSeq)

      val mergedChanges = changesPerDocument
        .groupMapReduce { case (documentUri, _) => documentUri } {
          case (_, edits) => edits.asScala.distinct
        }(_ ++ _)
        .mapValues(_.asJava) // deduplicate same changes
        .toMap
        .asJava
      new l.WorkspaceEdit(mergedChanges)
```

Reply:
The thing is that a single `WorkspaceEdit` can contain changes to the whole workspace and is not a change per document. Also, deduplication logic is missing from the suggested change